mav_trajectory_generation/test/test_polynomial_optimization.cpp

```diff
@@ -775,6 +773,69 @@ TEST(MavTrajectoryGeneration, 2_vertices_rand) {
   }
 }
 
+// Test unpacking and repacking constraints between [d_f; d_p] and p.
+// Test 2 vertices setup
+TEST(MavTrajectoryGeneration, ConstraintPacking) {
+  const int kMaxDerivative = derivative_order::JERK;
+  const size_t kNumSegments = 5;
+  Eigen::VectorXd min_pos, max_pos;
+  min_pos = Eigen::Vector3d::Constant(-50.0);
+  max_pos = -min_pos;
+  const int kSeed = 12345;
+  const size_t kNumSetups = 100;
+
+  for (size_t i = 0; i < kNumSetups; i++) {
+    Vertex::Vector vertices;
+    vertices = createRandomVertices(kMaxDerivative, kNumSegments, min_pos,
+                                    max_pos, kSeed + i);
+
+    const double kApproxVMax = 3.0;
+    const double kApproxAMax = 5.0;
+    std::vector<double> segment_times =
+        estimateSegmentTimes(vertices, kApproxVMax, kApproxAMax);
+
+    PolynomialOptimization<N> opt(3);
+    opt.setupFromVertices(vertices, segment_times);
+    opt.solveLinear();
+
+    Segment::Vector segments;
+    opt.getSegments(&segments);
+
+    std::vector<Eigen::VectorXd> fixed_constraints;
+    std::vector<Eigen::VectorXd> free_constraints;
+
+    opt.getFixedConstraints(&fixed_constraints);
+    opt.getFreeConstraints(&free_constraints);
+
+    // Get the mapping matrices.
+    Eigen::MatrixXd M, A_inv, A, M_pinv;
+    opt.getM(&M);
+    opt.getAInverse(&A_inv);
+    opt.getA(&A);
+    opt.getMpinv(&M_pinv);
+
+    ASSERT_EQ(fixed_constraints.size(), 3);
+    ASSERT_EQ(free_constraints.size(), 3);
+
+    // For each dimension... We can test that [df;dp] -> p -> [df;dp].
+    for (int i = 0; i < 3; ++i) {
+      Eigen::VectorXd d_all_ordered(fixed_constraints[i].size() +
+                                    free_constraints[i].size());
+      d_all_ordered << fixed_constraints[i], free_constraints[i];
+      Eigen::VectorXd p = A_inv * M * d_all_ordered;
+      Eigen::VectorXd d_unordered = A * p;
+      Eigen::VectorXd d_reordered = M_pinv * d_unordered;
+      EXPECT_TRUE(EIGEN_MATRIX_NEAR(d_all_ordered, d_reordered, 1e-6));
```

Comment:
Yes. :) This checks that converting back and forth between the polynomial coefficients and the packed/unpacked constraints works correctly. The content of A_inv and M are are checked when repacking the ps from d fixed and d parameters and then comparing it with the ps stored in segments. A and M pinv are checked by taking the ps are unpacking them again into d_unoreded and d_reordered.

I check the content of A specifically against the MATLAB implementation in some tests in local planning, but really these two matrices are only ever used together to switch between parameter orderings.